Makefile

```diff
@@ -9,25 +9,21 @@ else
     BUILD_DATE ?= $(shell date "$(DATE_FMT)")
 endif
 
-ifndef CGO_CPPFLAGS
-    export CGO_CPPFLAGS := $(CPPFLAGS)
-endif
-ifndef CGO_CFLAGS
-    export CGO_CFLAGS := $(CFLAGS)
-endif
-ifndef CGO_LDFLAGS
-    export CGO_LDFLAGS := $(LDFLAGS)
-endif
-
-GO_LDFLAGS := -X github.com/cli/cli/internal/build.Version=$(GH_VERSION) $(GO_LDFLAGS)
-GO_LDFLAGS := -X github.com/cli/cli/internal/build.Date=$(BUILD_DATE) $(GO_LDFLAGS)
-ifdef GH_OAUTH_CLIENT_SECRET
-	GO_LDFLAGS := -X github.com/cli/cli/internal/authflow.oauthClientID=$(GH_OAUTH_CLIENT_ID) $(GO_LDFLAGS)
-	GO_LDFLAGS := -X github.com/cli/cli/internal/authflow.oauthClientSecret=$(GH_OAUTH_CLIENT_SECRET) $(GO_LDFLAGS)
-endif
+CGO_CPPFLAGS ?= ${CPPFLAGS}
+export CGO_CPPFLAGS
+CGO_CFLAGS ?= ${CFLAGS}
+export CGO_CFLAGS
+CGO_LDFLAGS ?= ${LDFLAGS}
+export CGO_LDFLAGS
+
+GO_LDFLAGS += \
```

Comment:
This (perhaps unintentionally) changes the behavior of GO_LDFLAGS when they are passed via the environment. Before, custom GO_LDFLAGS were appended at the end of `-X` ones added here in the Makefile. After your change, users are no longer able to override `github.com/cli/cli/internal/build.Version` via `GO_LDFLAGS`, since the appended one would always take precedence.

This isn't a huge deal, since setting `GH_VERSION` is the preferred way to set the version rather than specifying the ldflag. Still, I would advise not submitting "cleanup" or "refactor" contributions to the project since they often cause unintentional side-effects and add little or no value to the codebase.

Please rebase this branch so that GO_LDFLAGS changes are taken out. Thank you!

Reply:
Ah, I see. Thanks.

Reply:
I've added `override` to fix that.

Reply:
Now that I think about it that isn't enough.

Reply:
BTW, the previous code did allow the command line variable spec to override the assignment, it always happens unless you use the `override` keyword. Also I'm not aware of any portable construct for that other than using auxiliary variables as intermediaries.

Reply:
Now, what I'm going to do is to use `override` and `:=` and add `GO_LDFLAGS` at the end so that it matches the behavior you describe, as it seems sensible.

Reply:
(because it's clear the previous code didn't do what was intended)

Reply:
The original way our Makefile section for GO_LDFLAGS was set up definitely did allow inheriting GO_LDFLAGS from the environment, as demonstrated by:

```
$ rm -f bin/gh; GO_LDFLAGS=HELLO make
go build -trimpath -ldflags "-X github.com/cli/cli/internal/build.Date=2020-11-23 -X github.com/cli/cli/internal/build.Version=v1.2.1-32-ge36c902 HELLO" -o "bin/gh" ./cmd/gh
```

I appreciate you trying to get the new version to work via `override`, but it I find it less easier to read or to understand compared to the original version. Therefore, I would suggesting leaving in our original version. Thank you!

Reply:
It however doesn't work when used as a parameter, like: `make GO_LDFLAGS=HELLO`.

Reply:
Or was that intended?

Reply:
@ismaell My apologies, I did not notice that `make GO_LDFLAGS=HELLO` and `GO_LDFLAGS=HELLO make` had different behaviors. I have a habit of always using the latter syntax.

Do you think adding an `override` directive to the first GO_LDFLAGS line would fix this? Then, I would ask you to make that change please. However, the rest of the GO_LDFLAGS setup could stay the same. 🙇

Reply:
Yes, `override` makes the behavior homogeneous, but it's all about expectations.

Conventionally, most build systems override the environment but not command line, so `VAR=... make` doesn't work, but `make VAR=...` does. IMHO, being always explicit, this is more reliable.

Instead of prepending to an existing variable, the normal approach is to define levels:
* Buildsystem-wide (e.g. `G_GO_LDFLAGS`)
* Target-specific (e.g. `bin/gh_GO_LDFLAGS`)
* User-provided (`GO_LDFLAGS`)

Buildsystem-wide and target-specific ones aren't meant for configuration, it's more for system-dependent flags, so the buildsystem is free to do as it pleases, the user-provided flags OTOH shouldn't be touched by the buildsystem if specified by the user, so `override` should be avoided. No composition is needed, as normally the rules would take all three.

So the rules look like:
```make
bin/gh: $(BUILD_FILES)
        go build -trimpath -ldflags "${G_GO_LDFLAGS} ${$@_GO_LDFLAGS} ${GO_LDFLAGS}" -o "$@" ./cmd/${@F}
```

Handling becomes much simpler.

Reply:
@ismaell Thank you for explaining! I do agree that several different types of variables would make it easier to compose a more sophisticated build config. But unless people who are building gh ask for this level of control, I do no think that we need to provide any extra features apart from ones that are currently supported.

Since you seem to have deep knowledge about Makefiles, I will leave it up to you to decide whether we would benefit from an `override` keyword here, or whether that's an anti-pattern. Overall, the changes here look good. 👍

Reply:
Probably the best option is to go with three variables, as that's the most portable.

Would there be interest to make it work with other Make implementations besides GNU's in the long term?

If not, then `override` would be fine.

Reply:
If the `override` declaration decreases portability, then I would vote to leave it out. For now, users can pass GO_LDFLAGS via environment if they want to add additional ldflags, and GO_LDFLAGS via argument to `make` if they want to completely replace the default ones.

We don't have plans to support non-GNU make, and we haven't received such requests yet, so we'll see.

> Probably the best option is to go with three variables, as that's the most portable.

That sounds true, but I feel that that increases the amount of possible `make` inputs to an amount that is unbecoming of a simple script that merely invokes `go build` to produce one binary. So for now, let's keep everything as-is and see if the needs of our users become more specific. Thanks for all your input so far! 🙇